File: 2025-12-18_week03/B/solutions/solution.py
def op1(lst, l, r, x):
    for i  in range(l - 1, r):
        lst[i] += x
    return lst

def op2(lst, l, r, x):
    for i in range(l -1, r):
        lst[i] -= x
    return lst

def op3(lst, l, r):
    return (max(lst[l - 1 : r]) - min(lst[l - 1 : r]))

n, q = map(int, input().split())

lst = [int(ele) for ele in input().split()]

ans = []

while(q > 0):
    opr = [int(ele) for ele in input().split()]
    
    if opr[0] == 1:
        lst = op1(lst, opr[1], opr[2], opr[3])

    elif opr[0] == 2:
        lst = op2(lst, opr[1], opr[2], opr[3])

    elif opr[0] == 3:
        rst = op3(lst, opr[1], opr[2])
        ans.append(rst)

    q -= 1


for i in ans:
    print(i)

File: 2025-12-18_week03/B/solutions/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    vector<long long> lst(n);
    for (int i = 0; i < n; i++) {
        cin >> lst[i];
    }

    vector<long long> ans;

    while (q--) {
        int op;
        cin >> op;

        if (op == 1) {
            int l, r;
            long long x;
            cin >> l >> r >> x;
            for (int i = l - 1; i < r; i++) {
                lst[i] += x;
            }
        }
        else if (op == 2) {
            int l, r;
            long long x;
            cin >> l >> r >> x;
            for (int i = l - 1; i < r; i++) {
                lst[i] -= x;
            }
        }
        else if (op == 3) {
            int l, r;
            cin >> l >> r;
            long long mx = lst[l - 1];
            long long mn = lst[l - 1];
            for (int i = l - 1; i < r; i++) {
                mx = max(mx, lst[i]);
                mn = min(mn, lst[i]);
            }
            ans.push_back(mx - mn);
        }
    }

    for (long long x : ans) {
        cout << x << '\n';
    }

    return 0;
}
